File: xposed-stub/src/main/java/de/robv/android/xposed/XC_MethodHook.java
package de.robv.android.xposed;

import java.lang.reflect.Member;

public abstract class XC_MethodHook {

    protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
    }

    protected void afterHookedMethod(MethodHookParam param) throws Throwable {
    }

    public static class MethodHookParam {
        public Member method;
        public Object thisObject;
        public Object[] args = new Object[0];
        public boolean returnEarly;

        private Object result;
        private Throwable throwable;

        public Object getResult() {
            return result;
        }

        public void setResult(Object result) {
            this.result = result;
            this.throwable = null;
            this.returnEarly = true;
        }

        public Throwable getThrowable() {
            return throwable;
        }

        public boolean hasThrowable() {
            return throwable != null;
        }

        public void setThrowable(Throwable throwable) {
            this.throwable = throwable;
            this.result = null;
            this.returnEarly = true;
        }

        public Object getResultOrThrowable() throws Throwable {
            if (throwable != null) {
                throw throwable;
            }
            return result;
        }
    }

    public class Unhook {
        private final Member hookedMethod;

        public Unhook(Member hookedMethod) {
            this.hookedMethod = hookedMethod;
        }

        public Member getHookedMethod() {
            return hookedMethod;
        }

        public void unhook() {
            // no-op in compile-time stub
        }
    }
}

File: app/src/main/java/io/github/magisk317/relay/xp/compat/XC_MethodHook.kt
package io.github.magisk317.relay.xp.compat

import java.lang.reflect.Member

abstract class XC_MethodHook {

    @Throws(Throwable::class)
    protected open fun beforeHookedMethod(param: MethodHookParam) {
    }

    @Throws(Throwable::class)
    protected open fun afterHookedMethod(param: MethodHookParam) {
    }

    internal fun dispatchBefore(param: MethodHookParam) {
        beforeHookedMethod(param)
    }

    internal fun dispatchAfter(param: MethodHookParam) {
        afterHookedMethod(param)
    }

    class MethodHookParam {
        lateinit var method: Member
        var thisObject: Any = Any()
        var args: Array<Any?> = emptyArray()
        var returnEarly: Boolean = false

        private var _result: Any? = null
        private var _throwable: Throwable? = null

        fun getResult(): Any? = _result

        fun setResult(value: Any?) {
            _result = value
            _throwable = null
            returnEarly = true
        }

        fun getThrowable(): Throwable? = _throwable

        fun setThrowable(value: Throwable?) {
            _throwable = value
            _result = null
            returnEarly = true
        }

        fun hasThrowable(): Boolean = _throwable != null

        @Throws(Throwable::class)
        fun getResultOrThrowable(): Any? {
            _throwable?.let { throw it }
            return _result
        }
    }

    class Unhook(
        private val hookedMethod: Member,
        private val action: (() -> Unit)? = null,
    ) {
        fun getHookedMethod(): Member = hookedMethod

        fun unhook() {
            action?.invoke()
        }
    }
}
